README.md
Overview
--------

This project provides a consistent build environment used for all releases
of GafferHQ/gaffer and GafferHQ/dependencies. It is not necessary to use this
environment to make your own builds, but it may be useful as a reference.

Prerequisites
-------------

- ARNOLD_ROOT environment variable pointing to installation of Arnold 5
- DELIGHT environment variable pointing to installation of 3delight 13
- Docker 17.05 or later

Usage
-----

Build and upload a gaffer release :

`./build.py --version 0.45.0.0 --upload 1`

Build and upload a dependencies release :

`./build.py --project dependencies --version 0.45.0.0 --upload 1`

Make a Linux release using Docker on a Mac :

`./build.py --docker 1 --arnoldRoot /path/to/linux/arnoldRoot --delightRoot /path/to/linux/delightRoot --version 0.45.0.0 --upload 1`

At times, builds may fail for assorted reasons. In such cases, the `-resume 1` argument can be used to attempt to continue making use of any existing build artifacts. Without this flag, the each build will start with a clean source tree/container.

Steps remaining to be automated
-------------------------------

- Update https://github.com/GafferHQ/documentation to include the latest docs.

Docker Cheatsheet
-----------------

Remove stopped containers and dangling images :

`docker system prune`

Remove all images, not just temporary container images :

`docker system prune -a`

Debug a stopped build container :

```
# Find the container of interest
docker ps -a
# Make a new image from container
docker commit <containerID> <imageName>
# Run interactive session in new image
docker run -it <imageName> /bin/bash
```

Building the build environment
------------------------------

Gaffer and Dependency builds are made using pre-published docker images. These
environments are built from the Dockerfile in this repository. The
`build-docker.py` script also aids the build/publish of these images. For
example:

 `./build-docker.py --tag 1.1.0 --upload`

> NOTE: The `build-docker.py` uses the simpler flag syntax for most of it options,
> this means its a little different to `build.py`, we aim to update that over
> time to match.

In order to produce this image, we make heavy use of `yum` to install the
packages required to build Gaffer and it's dependencies. As we don't manually
specify every version of every package we risk a non-deterministic build
environment. To get around this, we make use of `yum versionlock`. The
`yum-versionlock.list` file in the repository is copied into the base Centos
image such that when `yum` runs, it will repeatably install the expected
versions. In order to help manage this, the `build.py` script has a few options
to aid updating of the lock list when new packages are added or updates are required.

 - `--update-version-locks` When set this will ignore all version locks and
   update `yum-versionlock.list` to the 'current' version of all packages
   installed during docker's build. The revised file can then be committed and tagged,
   and a new docker image pushed to docker hub.

 - `--new-only` When set the existing version lock list will not be cleared.
   This allows the versions to be locked for any new packages installed by
   changes to the `Dockerfile` without affecting the versions of existing
   packages.

 - `--upload` Will push the built image to a docker hub tag.

> NOTE: GitHub Packages. We have an Action that will build the Docker image and publish
> it to GitHub Packages for the repo whenever a release is made. This image is perfectly
> usable, however, at the time of going to press, it requires a login to pull. We're
> maintaining the DockerHub hosted images until we have migrated CI/they fix this.

### Cheat sheet

Added a new package to the image, installed with yum in the Dockerfile:

`./build-docker.py --update-version-locks --new-only --tag x.x.x`

Update all packages to latest:

`./build-docker.py --update-version-locks --tag x.x.x`

### A note on Docker's caching mechanism

Docker caches layers based on the `RUN` command string. As such, it does not
known when the version lock file changes. the `--update-version-locks` command
will always run with `--no-cache` set, but if you've just pulled some updates
from upstream, and are re-building on your machine, you may find docker will
be using an out-of-date cache of one or more of the layers.

As such, its recommended to use `--no-cache` whenever performing release
builds.



Changes.md
1.2.0
=====

- Dockerfile :
  - Switched SCL packages to the `7.6.1810` vault.
  - Updated sundry SCL dependency versions.
  - Added automatic `pip` upgrade to latest supported version.
  - Added `libxkb-common-x11-devel` package.
  - Updated `sphinx` to 1.8.1.

1.1.0
=====

- build.py :
  - Added support for building `GafferRenderman`.
  - Moved to GitHub Packages hosted Docker image.
- Dockerfile :
  - Added `git` to the build image.
  - Added `lz4` and `lz4-devel` to the image.
  - Updated package versions to latest.

 1.0.0
 =====

 - Initial release supporting version-locked container build with scl devtoolset-6.


build-local.py
#!/usr/bin/env python
##########################################################################
#
#  Copyright (c) 2018, Image Engine Design Inc. All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are
#  met:
#
#      * Redistributions of source code must retain the above
#        copyright notice, this list of conditions and the following
#        disclaimer.
#
#      * Redistributions in binary form must reproduce the above
#        copyright notice, this list of conditions and the following
#        disclaimer in the documentation and/or other materials provided with
#        the distribution.
#
#      * Neither the name of John Haddon nor the names of
#        any other contributors to this software may be used to endorse or
#        promote products derived from this software without specific prior
#        written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
##########################################################################

import os
import sys
import distutils.util
import uuid
import json
import shutil
import argparse
import subprocess
import multiprocessing

parser = argparse.ArgumentParser()

parser.add_argument(
	"--build-env-version",
	dest = "buildEnvVersion",
	default = "1.0.0",
	help = "The container image tag to use for docker builds."
)

parser.add_argument(
	"--build-env-image",
	dest = "buildEnvImage",
	default = "gafferhq/build",
	help = "The container image to use for docker builds."
)

parser.add_argument(
	"--arnoldRoot",
	default = os.environ.get( "ARNOLD_ROOT", "" ),
	help = "The root of an installation of Arnold 5. "
	       "Note that if cross-compiling a Linux build "
	       "using Docker on a Mac, this must point to "
	       "a Linux build of Arnold."
)

parser.add_argument(
	"--delightRoot",
	default = os.environ.get( "DELIGHT", "" ),
	help = "The root of an installation of 3Delight 13. "
	       "Note that if cross-compiling a Linux build "
	       "using Docker on a Mac, this must point to "
	       "a Linux build of 3Delight."
)

parser.add_argument(
	"--source",
	default = os.getcwd(),
	help = "The root of the gaffer source tree"
)

parser.add_argument(
	"--buildDir",
	default = os.path.join( os.getcwd(), 'build' ),
	help = "The build directory, this should contain the Gaffer dependencies"
)

parser.add_argument(
	"--buildCacheDir",
	default = "",
	help = "A directory to use as the scons cache"
)

parser.add_argument(
	"--sconsCmd",
	default = "docs",
	help = "The scons command to call"
)

parser.add_argument(
	"--docker",
	type = distutils.util.strtobool,
	default = "linux" in sys.platform,
	help = "Performs the build using a Docker container. This provides a "
	       "known build platform so that builds are repeatable."
)

args = parser.parse_args()

# Check that the paths to the renderers are sane.

platform = "linux" if "linux" in sys.platform or args.docker else "osx"
libExtension = ".so" if platform == "linux" else ".dylib"

if args.arnoldRoot :
	arnoldLib = args.arnoldRoot + "/bin/libai" + libExtension
	if not os.path.exists( arnoldLib ) :
		parser.exit( 1, "{0} not found\n".format( arnoldLib ) )

if args.delightRoot :
	delightLib = args.delightRoot + "/lib/lib3delight" + libExtension
	if not os.path.exists( delightLib ) :
		parser.exit( 1, "{0} not found\n".format( delightLib ) )

# Build a little dictionary of variables we'll need over and over again
# in string formatting operations, and use it to figure out what
# package we will eventually be generating.

formatVariables = {
	"platform" : platform,
	"arnoldRoot" : args.arnoldRoot,
	"delight" : args.delightRoot,
	"buildDir" : args.buildDir,
	"buildCacheDir" : args.buildCacheDir,
	"source" : args.source
}

# Restart ourselves inside a Docker container so that we use a repeatable
# build environment.
if args.docker and not os.path.exists( "/.dockerenv" ) :

	image = "%s:%s" % ( args.buildEnvImage, args.buildEnvVersion )
	containerName = "gafferhq-build-{id}".format( id = uuid.uuid1() )

	containerEnv = []
	containerMounts = []

	containerMounts.append( "-v %s:/scripts:Z" % os.path.dirname( __file__ ) )
	containerMounts.append( "-v %s:/source:Z" % args.source )
	containerMounts.append( "-v %s:/build:Z" % args.buildDir )
	if args.buildCacheDir :
		containerMounts.append( "-v %s:/buildCache:Z" % args.buildCacheDir )

	if args.arnoldRoot :
		containerMounts.append( "-v %s:/arnold:ro,Z" % args.arnoldRoot )
		containerEnv.append( "ARNOLD_ROOT=/arnold" )

	if args.delightRoot :
		containerMounts.append( " -v %s:/delight:ro,Z" % args.delightRoot )
		containerEnv.append( "DELIGHT=/delight" )

	containerEnv = " ".join( containerEnv )
	containerMounts = " ".join( containerMounts )

	containerCommand = "env {env} bash -c '/scripts/build-local.py --sconsCmd={cmd} --buildDir=/build --buildCacheDir=/buildCache --source=/source'".format( env = containerEnv, cmd = args.sconsCmd, **formatVariables )

	dockerCommand = "docker run -it {mounts} --name {name} {image} {command}".format(
		source = args.source,
		buildDir = args.buildDir,
		mounts = containerMounts,
		name = containerName,
		image = image,
		command = containerCommand
	)
	sys.stderr.write( dockerCommand + "\n" )
	subprocess.check_call( dockerCommand, shell = True )

	sys.exit( 0 )

# Here we're actually doing the build, this will run either locally or inside
# the container bootstrapped above

if os.path.exists( "/.dockerenv" ) :

	# Start an X server so we can generate screenshots when the
	# documentation builds.
	os.system( "Xvfb :99 -screen 0 1280x1024x24 &" )
	os.environ["DISPLAY"] = ":99"
	os.system( "metacity&" )

os.chdir( args.source )

# Perform the build.

# We run SCons indirectly via `python` so that it uses our
# preferred python from the environment. SCons itself
# unfortunately hardcodes `/usr/bin/python`, which might not
# have the modules we need to build the docs.
buildCommand = "python `which scons` {cmd} ENV_VARS_TO_IMPORT=PATH DELIGHT_ROOT={delight} ARNOLD_ROOT={arnoldRoot} BUILD_DIR={buildDir} BUILD_CACHEDIR={buildCacheDir} OPTIONS='' -j {cpus}".format(
	cmd=args.sconsCmd,
	cpus=multiprocessing.cpu_count(),
	**formatVariables
)

sys.stderr.write( buildCommand + "\n" )
subprocess.check_call( buildCommand, shell=True )



versionlock.sh
#!/usr/bin/env bash

# A simple script to help manage yum versionlock lists.
#
# versionlock.sh list-installed /tmp/basePackages
#    Write a package-per-line list of installed packages to the supplied path
#
# versionlock.sh lock-new /tmp/basePackages
#    Add a yum versionlock for any packages installed now, that aren't listed
#    in the supplied package list (generated by list-installed earlier).
#

function dumpInstalled {
	# The 'tr' fun is to get around the fact that yum list always wraps lines. This puts them back.
	yum -q list installed | tr "\n" "#" | sed -e 's/# / /g' | tr "#" "\n" | cut -d ' ' -f 1 | sort > $1
}

if [ "$1" == "list-installed" ]; then

	dumpInstalled $2

elif [ "$1" == "lock-new" ]; then

	dumpInstalled "$2-delta"

	# This may have line too long issues at some point, but doing a lock-per
	# package is not only horrendously slow, but adds a timestamp for each
	# package which makes diffs harder to understand.
	yum versionlock $(diff --new-line-format="%L" --old-line-format="" --unchanged-line-format="" $2 "$2-delta" )

fi



build-docker.py
#!/usr/bin/env python
##########################################################################
#
#  Copyright (c) 2019, Cinesite VFX Ltd. All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are
#  met:
#
#      * Redistributions of source code must retain the above
#        copyright notice, this list of conditions and the following
#        disclaimer.
#
#      * Redistributions in binary form must reproduce the above
#        copyright notice, this list of conditions and the following
#        disclaimer in the documentation and/or other materials provided with
#        the distribution.
#
#      * Neither the name of John Haddon nor the names of
#        any other contributors to this software may be used to endorse or
#        promote products derived from this software without specific prior
#        written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
##########################################################################

import os
import sys
import distutils.util
import uuid
import argparse
import subprocess

parser = argparse.ArgumentParser(

	description =
	    "Builds the gafferhq container for linux builds. \n"
	    "Incorporates a version-locking mechanism to allow stable builds that "
	    "include yum-installed packages.",

	epilog =
	    "The yum version locking mechanism works as follows:\n"
	    " 1. The source checkout contains a yum-versionlock formatted list of\n"
	    "    package versions.\n"
	    " 2. The Dockerfile copies this in at the beginning of a build. \n"
	    " 3. Any yum installs during the build will then respect these versions.\n"
	    " 4. At the end of the build, if requested, we update the file with the\n"
	    "    versions of any new (or optionally - all) packages.\n"
	    " 5. The updated version list can be committed to source control.\n\n"
	    " NOTE: pip installed packages should be locked by including explicit "
	    " version numbers in their install commands."
)

parser.add_argument(
	"--upload",
	action = 'store_true',
	help = "Pushes the build image to dockerhub. You must first have used "
	       "'docker login' to set your credentials"
)

parser.add_argument(
	"--tag",
	dest = "tag",
	default = "latest",
	help = "The tag to apply to the build docker image when building the 'build' project."
)

parser.add_argument(
	"--image",
	default = "gafferhq/build",
	help = "The dockerhub organisation and image name to use in the image tag."
)

parser.add_argument(
	"--no-cache",
	dest = "noCache",
	action = 'store_true',
	help = "Because docker caches layers based on the RUN command string it "
	       "will fail to include any changes to package version locks, etc... "
	       "Use this flag to ensure a clean build."
)

parser.add_argument(
	"--update-version-locks",
	dest = "updateVersions",
	action = 'store_true',
	help = "If set, will remove the package manager version locks prior to "
	       "building the docker image to allow updated packages to be retrieved."
)

parser.add_argument(
	"--new-only",
	dest = "newPackagesOnly",
	action = 'store_true',
	help = "If set along with --update-version-locks, only newly added packages "
	       "without an existing version lock will be updated in the lock list, "
	       "existing packages will keep their version."
)

parser.add_argument(
	"--version-lock-file",
	dest = "versionlockFile",
	default = "yum-versionlock.list",
	help = "The local file to use for the yum versionlock mechanism."
)

args = parser.parse_args()

# Rough order of operations:
#
#  1. Copy in versionlock list from source control or a blank one
#  2. Perform build
#  3. Copy out versionlock list if requested
#  4. Upload final image to dockerhub

imageTag = "{image}:{tag}".format( image = args.image, tag = args.tag )

# 1. Versionlock management

# As we can't have conditionals in the Dockerfile, we always copy in our
# lock file, so to 'unlock' we simply empty it out so it has no effect.
if args.updateVersions and not args.newPackagesOnly :
	sys.stderr.write( "Unlocking all package versions...\n" )
	if os.path.exists( args.versionlockFile ) :
		os.remove( args.versionlockFile )
	open( args.versionlockFile, 'w' ).close()

# 2. Build the image

buildCommand = "docker build {cache} -t {tag} .".format(
	cache = "--no-cache" if ( args.noCache or args.updateVersions ) else "",
	tag = imageTag
)
sys.stderr.write( buildCommand + "\n" )
subprocess.check_call( buildCommand, shell = True )

# 3. Extract updated versionlock file

if args.updateVersions :
	# Extract the updated version lock files from the container
	extractCommand = " && ".join((
		"docker create --name {name} {tag}",
		"docker cp {name}:{versionlockSrc} {versionlockDest}",
		"docker rm {name}",
		"sort {versionlockDest} -o {versionlockDest}"
	)).format(
		name = "gafferhq-build-{id}".format( id = uuid.uuid1() ),
		tag = imageTag,
		versionlockSrc = "/etc/yum/pluginconf.d/versionlock.list",
		versionlockDest = args.versionlockFile
	)
	sys.stderr.write( extractCommand + "\n" )
	subprocess.check_call( extractCommand, shell = True )

# 4. Upload

if args.upload :
	pushCommand = "docker push %s" % imageTag
	sys.stderr.write( pushCommand + "\n" )
	subprocess.check_call( pushCommand, shell = True )


build.py
#!/usr/bin/env python
##########################################################################
#
#  Copyright (c) 2018, Image Engine Design Inc. All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are
#  met:
#
#      * Redistributions of source code must retain the above
#        copyright notice, this list of conditions and the following
#        disclaimer.
#
#      * Redistributions in binary form must reproduce the above
#        copyright notice, this list of conditions and the following
#        disclaimer in the documentation and/or other materials provided with
#        the distribution.
#
#      * Neither the name of John Haddon nor the names of
#        any other contributors to this software may be used to endorse or
#        promote products derived from this software without specific prior
#        written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
##########################################################################

import os
import sys
import distutils.util
import uuid
import json
import shutil
import argparse
import subprocess
import multiprocessing

parser = argparse.ArgumentParser()

parser.add_argument(
	"--build-env-version",
	dest = "buildEnvVersion",
	default = "1.2.0",
	help = "The container image tag to use for docker builds."
)

parser.add_argument(
	"--build-env-image",
	dest = "buildEnvImage",
	default = "gafferhq/build",
	help = "The container image to use for docker builds."
)

parser.add_argument(
	"--organisation",
	default = "GafferHQ",
	help = "The GitHub organisation containing the project to build."
)

parser.add_argument(
	"--project",
	default = "gaffer",
	choices = [ "gaffer", "dependencies" ],
	help = "The project to build."
)

parser.add_argument(
	"--arnoldRoot",
	default = os.environ.get( "ARNOLD_ROOT", "" ),
	help = "The root of an installation of Arnold 6. "
	       "Note that if cross-compiling a Linux build "
	       "using Docker on a Mac, this must point to "
	       "a Linux build of Arnold."
)

parser.add_argument(
	"--delightRoot",
	default = os.environ.get( "DELIGHT", "" ),
	help = "The root of an installation of 3Delight 13. "
	       "Note that if cross-compiling a Linux build "
	       "using Docker on a Mac, this must point to "
	       "a Linux build of 3Delight."
)

parser.add_argument(
	"--renderManRoot",
	default = os.environ.get( "RMANTREE", "" ),
	help = "The root of an installation of RenderMan 22.6 or later. "
	       "Note that if cross-compiling a Linux build "
	       "using Docker on a Mac, this must point to "
	       "a Linux build of RenderMan."
)

parser.add_argument(
	"--version",
	help = "The version to build. Can either be a tag or SHA1 commit hash."
)

parser.add_argument(
	"--upload",
	type = distutils.util.strtobool,
	default = "0",
	help = "Uploads the resulting package to the GitHub release page. You must "
	       "have manually created the release and release notes already."
)

parser.add_argument(
	"--docker",
	type = distutils.util.strtobool,
	default = "linux" in sys.platform,
	help = "Performs the build using a Docker container. This provides a "
	       "known build platform so that builds are repeatable."
)

parser.add_argument(
	"--interactive",
	type = distutils.util.strtobool,
	default = False,
	help = "When using docker, starts an interactive shell rather than "
		   "performing the build. This is useful for debugging."
)

args = parser.parse_args()

if args.interactive :
	if not args.docker :
		parser.exit( 1, "--interactive requires --docker\n" )
	if args.version or args.upload :
		parser.exit( 1, "--interactive can not be used with other flags\n" )
else :
	if not args.version :
		parser.exit( "--version argument is required")

# Check that our environment contains everything we need to do a build.

if args.upload :

	if "GITHUB_RELEASE_TOKEN" not in os.environ	:
		parser.exit( 1,  "GITHUB_RELEASE_TOKEN environment variable not set\n" )

	if not args.arnoldRoot :
		parser.exit( 1,  "Release builds must include Arnold (set $ARNOLD_ROOT or --arnoldRoot)\n" )
	if not args.delightRoot :
		parser.exit( 1,  "Release builds must include 3Delight (set $DELIGHT_ROOT or --delightRoot)\n" )


# Check that the paths to the renderers are sane.

platform = "linux" if "linux" in sys.platform or args.docker else "osx"
libExtension = ".so" if platform == "linux" else ".dylib"

if args.arnoldRoot :
	arnoldLib = args.arnoldRoot + "/bin/libai" + libExtension
	if not os.path.exists( arnoldLib ) :
		parser.exit( 1, "{0} not found\n".format( arnoldLib ) )

if args.delightRoot :
	delightLib = args.delightRoot + "/lib/lib3delight" + libExtension
	if not os.path.exists( delightLib ) :
		parser.exit( 1, "{0} not found\n".format( delightLib ) )

if args.renderManRoot :
	renderManLib = args.renderManRoot + "/lib/libprman" + libExtension
	if not os.path.exists( renderManLib ) :
		parser.exit( 1, "{0} not found\n".format( renderManLib ) )

# Build a little dictionary of variables we'll need over and over again
# in string formatting operations, and use it to figure out what
# package we will eventually be generating.

formatVariables = {
	"organisation" : args.organisation,
	"project" : args.project,
	"version" : args.version,
	"upload" : args.upload,
	"platform" : platform,
	"arnoldRoot" : args.arnoldRoot,
	"delight" : args.delightRoot,
	"renderManRoot" : args.renderManRoot,
	"releaseToken" : "",
	"auth" : "",
}

githubToken = os.environ.get( "GITHUB_RELEASE_TOKEN", "" )
if githubToken :
	formatVariables[ "releaseToken" ] = githubToken
	formatVariables[ "auth" ] = '-H "Authorization: token %s"' % githubToken

if args.project == "gaffer" :
	formatVariables["uploadFile"] = "{project}-{version}-{platform}.tar.gz".format( **formatVariables )
else :
	formatVariables["uploadFile"] = "gafferDependencies-{version}-{platform}.tar.gz".format( **formatVariables )

# If we're going to be doing an upload, then check that the release exists. Better
# to find out now than at the end of a lengthy build.

def releaseId() :

	release = subprocess.check_output(
		"curl -s {auth} https://api.github.com/repos/{organisation}/{project}/releases/tags/{version}".format(
			**formatVariables
		),
		shell = True
	)
	release = json.loads( release )
	return release.get( "id" )

if args.upload and releaseId() is None :
	parser.exit( 1, "Release {version} not found\n".format( **formatVariables ) )

# Restart ourselves inside a Docker container so that we use a repeatable
# build environment.
if args.docker and not os.path.exists( "/.dockerenv" ) :

	image = "%s:%s" % ( args.buildEnvImage, args.buildEnvVersion )
	containerName = "gafferhq-build-{id}".format( id = uuid.uuid1() )

	# We don't keep build.py in the images (otherwise we'd have to maintain
	# backwards compatibility when changing this script), so copy it in

	containerPrepCommand = " && ".join( (
		"docker create --name {name} {image}",
		"docker cp build.py {name}:/build.py",
		# This saves our changes to that container, so we can pick it up
		# in run later. We can't use exec as when you 'start' the image
		# it immediately exits as there is nothing to do. Docker is process
		# centric not 'machine' centric. You can either add in nasty sleep
		# commands into the image, but this seems to be the more 'docker'
		# way to do it.
		"docker commit {name} {image}-run",
		"docker rm {name}"
	) ).format(
		name = containerName,
		image = image
	)
	sys.stderr.write( containerPrepCommand + "\n" )
	subprocess.check_call( containerPrepCommand, shell = True )

	containerEnv = []
	if githubToken :
		containerEnv.append( "GITHUB_RELEASE_TOKEN=%s" % githubToken )

	containerMounts = []
	if args.arnoldRoot :
		containerMounts.append( "-v %s:/arnold:ro,Z" % args.arnoldRoot )
		containerEnv.append( "ARNOLD_ROOT=/arnold" )
	if args.delightRoot :
		containerMounts.append( " -v %s:/delight:ro,Z" % args.delightRoot )
		containerEnv.append( "DELIGHT=/delight" )
	if args.renderManRoot :
		containerMounts.append( "-v %s:/renderMan:ro,Z" % args.renderManRoot )
		containerEnv.append( "RMANTREE=/renderMan" )

	containerEnv = " ".join( containerEnv )
	containerMounts = " ".join( containerMounts )

	if args.interactive :
		containerCommand = "env {env} bash".format( env = containerEnv )
	else :
		containerCommand = "env {env} bash -c '/build.py --project {project} --version {version} --upload {upload}'".format( env = containerEnv, **formatVariables )

	dockerCommand = "docker run -it {mounts} --name {name} {image}-run {command}".format(
		mounts = containerMounts,
		name = containerName,
		image = image,
		command = containerCommand
	)
	sys.stderr.write( dockerCommand + "\n" )
	subprocess.check_call( dockerCommand, shell = True )

	if not args.interactive :
		# Copy out the generated package.
		copyCommand = "docker cp {container}:/{project}-{version}-source/{uploadFile} ./".format(
			container = containerName,
			**formatVariables
		)
		sys.stderr.write( copyCommand + "\n" )
		subprocess.check_call( copyCommand, shell = True )

	sys.exit( 0 )

# Here we're actually doing the build, this will run either locally or inside
# the container bootstrapped above

if os.path.exists( "/.dockerenv" ) and args.project == "gaffer" :

	# Start an X server so we can generate screenshots when the
	# documentation builds.
	os.system( "Xvfb :99 -screen 0 1280x1024x24 &" )
	os.environ["DISPLAY"] = ":99"
	os.system( "metacity&" )

# Download source code

sourceURL = "https://github.com/{organisation}/{project}/archive/{version}.tar.gz".format( **formatVariables )
sys.stderr.write( "Downloading source \"%s\"\n" % sourceURL )

sourceDirName = "{project}-{version}-source".format( **formatVariables )
tarFileName = "{0}.tar.gz".format( sourceDirName )
downloadCommand = "curl -L {0} > {1}".format( sourceURL, tarFileName )
sys.stderr.write( downloadCommand + "\n" )
subprocess.check_call( downloadCommand, shell = True )

sys.stderr.write( "Decompressing source to \"%s\"\n" % sourceDirName )

shutil.rmtree( sourceDirName, ignore_errors = True )
os.makedirs( sourceDirName )
subprocess.check_call( "tar xf %s -C %s --strip-components=1" % ( tarFileName, sourceDirName ), shell = True )
os.chdir( sourceDirName )

# Download precompiled dependencies. We do this using the
# same script that is used to download the dependencies for
# testing on Travis, so that release builds are always made
# against the same dependencies we have tested against.

if args.project == "gaffer" :
	# The scripts moved in #3242
	depsInstallScript = "./config/installDependencies.sh"
	if not os.path.exists( os.path.join( os.getcwd(), depsInstallScript ) ) :
		depsInstallScript = "./config/travis/installDependencies.sh"
	subprocess.check_call( depsInstallScript, shell = True )

# Perform the build.

if args.project == "gaffer" :

	# We run SCons indirectly via `python` so that it uses our
	# preferred python from the environment. SCons itself
	# unfortunately hardcodes `/usr/bin/python`, which might not
	# have the modules we need to build the docs.
	buildCommand = "python `which scons` package PACKAGE_FILE={uploadFile} ENV_VARS_TO_IMPORT=PATH DELIGHT_ROOT={delight} ARNOLD_ROOT={arnoldRoot} RENDERMAN_ROOT={renderManRoot} OPTIONS='' -j {cpus}".format(
		cpus=multiprocessing.cpu_count(), **formatVariables
	)

else :

	buildCommand = "env RMAN_ROOT={delight} ARNOLD_ROOT={arnoldRoot} ./build.py --buildDir {cwd}/gafferDependenciesBuild".format(
		cwd = os.getcwd(),
		**formatVariables
	)

sys.stderr.write( buildCommand + "\n" )
subprocess.check_call( buildCommand, shell=True )

# Upload the build

if args.upload :

	uploadCommand = (
		'curl {auth}'
		' -H "Content-Type: application/zip"'
		' --data-binary @{uploadFile} "{uploadURL}"'
		' -o /tmp/curlResult.txt' # Must specify output file in order to get progress output
	).format(
		uploadURL = "https://uploads.github.com/repos/{organisation}/{project}/releases/{id}/assets?name={uploadName}".format(
			id = releaseId(),
			uploadName = os.path.basename( formatVariables["uploadFile"] ),
			**formatVariables
		),
		**formatVariables
	)

	sys.stderr.write( "Uploading package\n" )
	sys.stderr.write( uploadCommand + "\n" )

	subprocess.check_call( uploadCommand, shell = True )
